_source/_posts/2022-10-01-spring-vault.md

```diff
@@ -0,0 +1,639 @@
+- Secrets management with HashiCorp's Vault
+- Using Spring Cloud Vault
+
+> **This tutorial was created with the following frameworks and tools**:
+> - [JHipster 7.9.3](https://www.jhipster.tech/installation/)
+> - [Java OpenJDK 17](https://jdk.java.net/java-se-ri/17)
+> - [Okta CLI 0.10.0](https://cli.okta.com)
+> - [Docker 20.10.12](https://docs.docker.com/engine/install/)
+> - [HTTPie 3.2.1](https://httpie.io/docs/cli/installation)
+> - [Vault 1.11.3](https://hub.docker.com/_/vault)
+
+{% include toc.md %}
+
+## Use Environment Variables for Secrets; a Precursor to Spring Vault
+
+Spring Boot applications can bind property values from environment variables. To demonstrate, create a `vault-demo-app` with OpenID Connect (OIDC) authentication, using the Spring Initializr. Then add `web`, `okta`, and `cloud-config-client` dependencies, some of which will be required later in the tutorial:
+
+```shell
+https start.spring.io/starter.zip \
+  bootVersion==2.7.4 \
+  dependencies==web,okta,cloud-config-client \
+  groupId==com.okta.developer \
+  artifactId==vault-demo-app  \
+  name=="Spring Boot Application" \
+  description=="Demo project of a Spring Boot application with Vault protected secrets" \
+  packageName==com.okta.developer.vault > vault-demo-app.zip
+```
+
+Unzip the file and open the project. Modify its `src/main/java/.../Application.java` class to add the `/` HTTP endpoint:
+
+```java
+package com.okta.developer.vault;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@SpringBootApplication
+public class Application {
+
+    public static void main(String[] args) {
+        SpringApplication.run(Application.class, args);
+    }
+
+    @GetMapping("/")
+    String hello(@AuthenticationPrincipal OidcUser user) {
+        return String.format("Welcome, %s", user.getFullName());
+    }
+
+}
+```
+
+Disable the cloud configuration for the first run. Edit `application.properties` and add the following value:
+
+```properties
+spring.cloud.config.enabled=false
+```
+
+### OpenID Connect authentication with Okta
+
+In a command line session, go to the `vault-demo-app` root folder.
+
+{% include setup/cli.md type="web" framework="Okta Spring Boot Starter" %}
+
+Instead of storing Okta credentials in `application.properties`, Spring Boot allows you to bind properties from environment variables. You can see this in action by starting your application with the Maven command below:
+
+```shell
+OKTA_OAUTH2_ISSUER={yourIssuerURI} \
+OKTA_OAUTH2_CLIENT_ID={yourClientId} \
+OKTA_OAUTH2_CLIENT_SECRET={yourClientSecret} \
+./mvnw spring-boot:run
+```
```

Comment:
I don't know if this part is necessary. The Okta CLI already configured the app and we're adding a painful step here. You have copy/paste the values when you can just run it and everything will work.

Reply:
But we don't want to store secrets in the project code. The section is about showing how to pass secrets through env. Please let me know if I miunderstood your comment.